File: cses/DP/python/minimize_coin.py


class Solution:

    def coin_change_rec(self, coins, index, target):
        if index == 0:
            if target % coins[index] == 0:
                return target // coins[index]
            return 1e9
        not_take = self.coin_change_rec(
            coins, index=index - 1, target=target) + 0
        take = 1e9
        if target - coins[index] >= 0:
            take = self.coin_change_rec(
                coins, index=index, target=target - coins[index]) + 1

        return min(take, not_take)

    def coin_change(self, coins, target):
        dp = [[0 for _ in range(target+1)] for _ in range(len(coins))]
        n = len(coins)
        for i in range(0, target+1):
            if i % coins[0] == 0:
                dp[0][i] = i // coins[0]
            else:
                dp[0][i] = float('inf')

        for c in range(1, len(coins)):
            for t in range(target+1):
                not_take = dp[c-1][t] + 0
                take = float('inf')
                if t - coins[c] >= 0:
                    take = dp[c][t - coins[c]] + 1
                dp[c][t] = min(take, not_take)
        return dp[n-1][target]


if __name__ == '__main__':
    s = Solution()
    n, target = list(map(int, input().split()))
    coins = list(map(int, input().split()))
    ans = s.coin_change(coins, target=target)
    if ans == float('inf'):
        print('-1')
    else:
        print(ans)

File: cses/DP/cpp/minimize_coin.cpp
#include <bits/stdc++.h>
#define int_max MOD + 2;
#define MOD 1000000007;
using namespace std;

class Solution
{
public:
    int coinChange(vector<int> &coins, int target)
    {
        int n = coins.size();
        vector<vector<int>> dp(n, vector<int>(target + 1, 0));

        for (int i = 0; i <= target; i++)
        {
            if (i % coins[0] == 0)
            {
                dp[0][i] = i / coins[0];
            }
            else
            {
                dp[0][i] = 1e9;
            }
        }

        for (int c = 1; c < n; c++)
        {
            for (int t = 0; t <= target; t++)
            {
                int not_take = dp[c - 1][t];
                int take = 1e9;
                if (t - coins[c] >= 0)
                {
                    take = dp[c][t - coins[c]] + 1;
                }
                dp[c][t] = min(take, not_take);
            }
        }

        return dp[n - 1][target] >= 1e9 ? -1 : dp[n - 1][target];
    }
};

int main()
{
    Solution s;
    int n, target;
    cin >> n >> target;
    vector<int> coins(n);
    for (int i = 0; i < n; i++)
    {
        cin >> coins[i];
    }

    int ans = s.coinChange(coins, target);
    if (ans == -1)
    {
        cout << "-1" << endl;
    }
    else
    {
        cout << ans << endl;
    }

    return 0;
}
